client/util.go

```diff
@@ -69,3 +70,84 @@ func pollingWatcher(ctx context.Context, client Client, group *key.Group, log lo
 
 	return ch
 }
+
+// newWatchAggregator maintains state of consumers calling `Watch` so that a
+// single `watch` request is made to the underlying client.
+func newWatchAggregator(c Client, l log.Logger) *watchAggregator {
+	return &watchAggregator{
+		Client:      c,
+		log:         l,
+		subscribers: make([]subscriber, 0),
+	}
+}
+
+type subscriber struct {
+	ctx context.Context
+	c   chan Result
+}
+
+type watchAggregator struct {
+	Client
+	log log.Logger
+
+	subscriberLock sync.Mutex
+	subscribers    []subscriber
+}
+
+func (c *watchAggregator) Watch(ctx context.Context) <-chan Result {
+	c.subscriberLock.Lock()
+	defer c.subscriberLock.Unlock()
+
+	sub := subscriber{ctx, make(chan Result, 5)}
+	c.subscribers = append(c.subscribers, sub)
+
+	if len(c.subscribers) == 1 {
+		ctx, cancel := context.WithCancel(context.Background())
+		go c.distribute(c.Client.Watch(ctx), cancel)
```

Comment:
Am i understanding it right that this go routine may be triggered depend on the outside's requests ? If so, it doesn't seem very safe to do so without bounding the number of go routines - We'd usually want a consumer/producer model to handle things like that no ?

Reply:
it'll spin up the background routine when number of subscribers goes from 0 to 1. the goroutine ends itself when subscribers goes from 1 to 0.

Reply:
Sorry I had my mind on the http server -_-" all good.